Add support for Performance / Efficency cores.

FractalFir opened the pull request:
(Hopefully) fixes: https://github.com/rust-lang/rustc-perf/issues/2008 on Linux.

Uses directories `/sys/devices/` to detect if the system has a "Performace-Hybrid" architecture(has P and E cores). If P cores are detected, forces the benchmark to only run on Performance cores using `taskset`.

Changes the P-core specific perf events `cpu_core/instructions:u/` to "normal" events  `instructions:u`.  I am a bit unsure if my implementation of that is correct, tough.

FractalFir commented:
It seems like this repo is using a fairly old version of Rust(` 1.75.0` which is 1 year old) in its CI. 

Is there a reason this can't be raised(eg. MSRV)? If so, I can adapt this patch to use older Rust.

Kobzol commented:
> It seems like this repo is using a fairly old version of Rust( 1.75.0 which is 1 year old) in its CI.

I like that Rust is bringing an "always up-to-date" philosophy, so that a year old version is considered to be "fairly old" :laughing:

I don't think that it's needed to stay on 1.75, feel free to update to a newer version, I'd say 1.82 the newest.

Kobzol commented:
I don't have P/E cores, so I can't test this, but I assume that it works for you. @bjorn3 Since you created https://github.com/rust-lang/rustc-perf/issues/2008, could you please test if this patch helps to make rustc-perf work on your P/E system?

FractalFir commented:
> I don't have P/E cores, so I can't test this, but I assume that it works for you.

I *think* it works for me(I can see the profiling results), but I also don't know enough about `rustc-perf` to conform that the numbers are accurate.

Kobzol commented:
Could you send some icount results for a few benchmarks? Just so that we can compare. And also the number of (P) cores that you have.

lqd commented:
Could you also rustfmt your code please, as well as fixing the unused warnings? (there seem to be a few typos in there as well). Thanks.

FractalFir commented:
It looks like some of the CI failures are caused by code I did not touch. I don' think fixing them in this PR would be right - I can fix them in another PR if needed. 

I *think* I caught all the typos in comments. 

As for the icounts, I only have the ones for an experimental patch to `rustc`. They are different from the ones on the website, but that might just be caused by the patch in question. I guess I'll have to run a benchmark on the baseline compiler. 

For summary-check they are(taken from the graph):
1.461 incr-full
1 full
0.887 incr-patched
0.387 incr-unchanged
For bitmaps-3.1.0-opt, they are roughly

10.81 G incr-full
7.64 G  full
3.21 G incr-patched
2.66 G incr-unchanged

I have 8 P cores(12 if you count the ones from hyper-threading) .

Kobzol commented:
Interesting. I have this:
![image](https://github.com/user-attachments/assets/9b1f61c8-246c-47d7-bc33-d898c0657cd3)

With the stable compiler (1.84.1), on a PC with 8 cores. The ratios are similar, but the absolute counts are quite different.

Kobzol commented:
https://github.com/rust-lang/rustc-perf/pull/2042

FractalFir commented on collector/src/compile/execute/mod.rs:
```diff
@@ -129,7 +129,55 @@ pub struct CargoProcess<'a> {
+	// Check if /sys/devices/cpu_core exists - if it does, then this is a "Performance-hybrid" CPU.
+	else if std::fs::exists("/sys/devices/cpu_core").expect("Could not check the CPU architecture detali: could not check if `/sys/devices/cpu_core` exists!") {
+		Some(std::fs::read_to_string("/sys/devices/cpu_core/cpus").unwrap().trim().to_string())
+	}else{
```
I don't know why it does not get formatted - I run `cargo fmt` 4 times now, and nothing seems to change.

nnethercote replied:
rustfmt sometimes silently fails to format parts of the code. It's hard to know why. In the past I have used trial and error to (temporarily) remove parts of the unformatted code until rustfmt starts working again, to determine what is blocking it. Last time I did that I found that a comment in the middle of a multi-line pattern was blocking rustfmt, see [this commit](https://github.com/rust-lang/rust/commit/7be593e379e4d5737700b1a09887992caea7cf87).

FractalFir commented on collector/src/compile/execute/mod.rs:
```diff
@@ -148,7 +196,12 @@ impl<'a> CargoProcess<'a> {
 
     fn base_command(&self, cwd: &Path, subcommand: &str) -> Command {
-        let mut cmd = Command::new(Path::new(&self.toolchain.components.cargo));
+        // Processors with P and E cores require special handling
```
The main reason "why" is that the E core perf results are incomplete/unreliable - I don't know _why_, exactly. 

I have considered printing something(a warning) in this case, but I was worried that would break scripts using `rustc-perf`.

Would setting CARGO_BUILD_JOBS to the core count be a good idea here?

Kobzol commented:
It would be better to rebase on top of master and then force-push, than to create the merge commit. You can check https://rustc-dev-guide.rust-lang.org/git.html#rebasing for some hints.

FractalFir commented:
I fixed the suggestions made here:
1. The entire thing is now properly rebased
2. Added some comments explaining *why* E cores are not used.
3. Added a warning message, informing the user that only P cores are used.
4. `CARGO_BUILD_JOBS` is now set to the P core count, to reflect the restriction to P only use P cores.

Kobzol reviewed (approved):
Thanks! I appreciate that you took the time to fix this issue.

Since rustfmt choked on your code, I becometh rustfmt in the latest commit.
